File: examples/SmartHomeDemo/src/User/SmartHome/EufyDoorbell.cs
using System;
using System.Collections.Generic;

namespace SmartHome;

public partial class EufyDoorbell
{
    private int _ringCount;
    private string _lastSnapshot = "Porch is quiet.";

    public EufyDoorbell(string deviceId, string displayName)
    {
        DeviceId = deviceId;
        DisplayName = displayName;
    }

    public HashSet<DeviceFeature> GetSupportedFeatures()
    {
        return
        [
            DeviceFeature.Camera,
            DeviceFeature.Motion,
            DeviceFeature.Doorbell,
            DeviceFeature.Automation,
        ];
    }

    public string DescribeState()
    {
        return $"motion={(MotionDetected ? "detected" : "clear")}, rings={_ringCount}, snapshot=\"{_lastSnapshot}\"";
    }

    public string CaptureSnapshot()
    {
        MotionDetected = true;
        _lastSnapshot = "Package courier visible at the porch.";
        return _lastSnapshot;
    }

    public void Ring()
    {
        _ringCount++;
        MotionDetected = true;
        _lastSnapshot = "Visitor pressed the porch doorbell.";
    }
}

File: examples/SmartHomeDemo/cpp/user/src/SmartHome/EufyDoorbell.cpp
#include "SmartHome/EufyDoorbell.hpp"

#include <utility>

namespace SmartHome
{

EufyDoorbell::EufyDoorbell(::Crimson::Cpp::String deviceId, ::Crimson::Cpp::String displayName)
{
    SetDeviceId(std::move(deviceId));
    SetDisplayName(std::move(displayName));
}

::Crimson::Cpp::Set<DeviceFeature> EufyDoorbell::GetSupportedFeatures()
{
    return
    {
        DeviceFeature::Camera,
        DeviceFeature::Motion,
        DeviceFeature::Doorbell,
        DeviceFeature::Automation,
    };
}

::Crimson::Cpp::String EufyDoorbell::DescribeState()
{
    return "motion=" + ::Crimson::Cpp::String(GetMotionDetected() ? "detected" : "clear")
        + ", rings=" + std::to_string(ringCount_)
        + ", snapshot=\"" + lastSnapshot_ + "\"";
}

::Crimson::Cpp::String EufyDoorbell::CaptureSnapshot()
{
    SetMotionDetected(true);
    lastSnapshot_ = "Package courier visible at the porch.";
    return lastSnapshot_;
}

void EufyDoorbell::Ring()
{
    ++ringCount_;
    SetMotionDetected(true);
    lastSnapshot_ = "Visitor pressed the porch doorbell.";
}

}
